src/result/result.py

```diff
@@ -482,3 +507,52 @@ def is_err(result: Result[T, E]) -> TypeGuard[Err[E]]:
     >>>     r   # r is of type Err[str]
     """
     return result.is_err()
+
+
+def do(gen: Generator[Result[T, E], None, None]) -> Result[T, E]:
+    """Do notation for Result (syntactic sugar for sequence of `and_then()` calls).
+
+
+    Usage:
+    >>> # This is similar to
+    use do_notation::m;
+    let final_result = m! {
+        x <- Ok("hello");
+        y <- Ok(True);
+        Ok(len(x) + int(y) + 0.5)
+    };
+
+    >>> final_result: Result[float, int] = do(
+            Ok(len(x) + int(y) + 0.5)
+            for x in Ok("hello")
+            for y in Ok(True)
+        )
+
+    NOTE: If you exclude the type annotation e.g. `Result[float, int]`
+    your type checker might be unable to infer the return type.
+    To avoid an error, you might need to help it with the type hint.
+    """
+    try:
+        return next(gen)
+    except DoException as e:
+        out: Err[E] = e.err  # type: ignore
+        return out
+
+
+if __name__ == "__main__":
+    import ast, macros
+
+    class RewriteXY(ast.NodeTransformer):
+        def visit_Name(self, node: ast.Name) -> ast.Name:
+            if node.id == 'x':
+                return ast.Name(id='y', ctx=node.ctx)
+            else:
+                return node
+
+    @macros.macro_processor(macros.STMT_MACRO, 0)
+    def replace_x_y(node: ast.AST) -> ast.AST:
+        return RewriteXY.visit(node)
+
+    # y = 123
+    # with replace_x_y!:
+    #     print(x)  # Rewritten to `print(y)`
```

Comment:
Sorry didn't notice this before. We need to remove this `if __name__ == '__main__":` block